### ISSUE TITLE
Rule proposal: no unchecked casts out of a potentially nullish type

### ISSUE THREAD

**Issue by legowerewolf** [typescript-eslint/typescript-eslint#7757]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

This rule should check that you're not casting away from a null or undefined value without checking it first.

### Fail Cases

```typescript
function t(param: string | undefined) {
    console.log(param as string);
}
```

### Pass Cases

```typescript
function t(param: string | undefined) {
    if (param === undefined) return;
    console.log(param);
}

// or

function t(param: string | undefined) {
    if (!param) return;
    console.log(param);
}
```

**Comment by Josh-Cena** [comment]
@bradzacher the template is indeed filled out?

**Comment by bradzacher** [comment]
@Josh-Cena one glaring issue remaining
![image](https://github.com/typescript-eslint/typescript-eslint/assets/7462525/dc680fe9-c0e6-4a60-9c78-e49840f99799)

**Comment by legowerewolf** [comment]
Oops. Missed the title.

**Comment by bradzacher** [comment]
in general we recommend banning `as` assertions due to their super unsafe bivarant nature.
This seems like a type-aware version of [`no-non-null-assertion`](https://typescript-eslint.io/rules/no-non-null-assertion).

There's a difficult thing here in that a lot of cases aren't going to be easy for us to handle thanks to [this missing API](https://github.com/microsoft/TypeScript/issues/9879). I guess we could build something that's pretty naive like "if left type includes `undefined` and/or `null` and the right type doesn't - error".

**Comment by legowerewolf** [comment]
Yeah, at work we've found that nearly every place there's an `as <type>` there's a problem *somewhere*. There's a lot of it, too.

**Comment by Josh-Cena** [comment]
Closing in favor of https://github.com/typescript-eslint/typescript-eslint/issues/7173—now that we have the assignability API, we should go directly for that rule!
